src/network/connectionthreads.cpp

```diff
@@ -0,0 +1,1402 @@
+	PROFILE(ThreadIdentifier << "ConnectionSend: [" << m_connection->getDesc() << "]");
+
+	/* if stop is requested don't stop immediately but try to send all        */
+	/* packets first */
+	while (!stopRequested() || packetsQueued()) {
+		BEGIN_DEBUG_EXCEPTION_HANDLER
+		PROFILE(ScopeProfiler sp(g_profiler, ThreadIdentifier.str(), SPT_AVG));
+
+		m_iteration_packets_avaialble = m_max_data_packets_per_iteration;
+
+		/* wait for trigger or timeout */
+		m_send_sleep_semaphore.wait(50);
+
+		/* remove all triggers */
+		while (m_send_sleep_semaphore.wait(0)) {
+		}
+
+		lasttime = curtime;
+		curtime = porting::getTimeMs();
+		float dtime = CALC_DTIME(lasttime, curtime);
+
+		/* first do all the reliable stuff */
+		runTimeouts(dtime);
+
+		/* translate commands to packets */
+		ConnectionCommand c = m_connection->m_command_queue.pop_frontNoEx(0);
+		while (c.type != CONNCMD_NONE) {
+			if (c.reliable)
+				processReliableCommand(c);
+			else
+				processNonReliableCommand(c);
+
+			c = m_connection->m_command_queue.pop_frontNoEx(0);
+		}
+
+		/* send non reliable packets */
+		sendPackets(dtime);
+
+		END_DEBUG_EXCEPTION_HANDLER
+	}
+
+	PROFILE(g_profiler->remove(ThreadIdentifier.str()));
+	return NULL;
+}
+
+void ConnectionSendThread::Trigger()
+{
+	m_send_sleep_semaphore.post();
+}
+
+bool ConnectionSendThread::packetsQueued()
+{
+	std::list<u16> peerIds = m_connection->getPeerIDs();
+
+	if (!m_outgoing_queue.empty() && !peerIds.empty())
+		return true;
+
+	for (u16 peerId : peerIds) {
+		PeerHelper peer = m_connection->getPeerNoEx(peerId);
+
+		if (!peer)
+			continue;
+
+		if (dynamic_cast<UDPPeer *>(&peer) == 0)
+			continue;
+
+		for (Channel &channel : (dynamic_cast<UDPPeer *>(&peer))->channels) {
+			if (!channel.queued_commands.empty()) {
+				return true;
+			}
+		}
+	}
+
+
+	return false;
+}
+
+void ConnectionSendThread::runTimeouts(float dtime)
+{
+	std::list<u16> timeouted_peers;
+	std::list<u16> peerIds = m_connection->getPeerIDs();
+
+	for (u16 &peerId : peerIds) {
+		PeerHelper peer = m_connection->getPeerNoEx(peerId);
+
+		if (!peer)
+			continue;
+
+		if (dynamic_cast<UDPPeer *>(&peer) == 0)
+			continue;
+
+		PROFILE(std::stringstream peerIdentifier);
+		PROFILE(peerIdentifier << "runTimeouts[" << m_connection->getDesc()
+			<< ";" << peerId << ";RELIABLE]");
+		PROFILE(ScopeProfiler
+		peerprofiler(g_profiler, peerIdentifier.str(), SPT_AVG));
+
+		SharedBuffer<u8> data(2); // data for sending ping, required here because of goto
+
+		/*
+			Check peer timeout
+		*/
+		if (peer->isTimedOut(m_timeout)) {
+			infostream << m_connection->getDesc()
+				<< "RunTimeouts(): Peer " << peer->id
+				<< " has timed out."
+				<< " (source=peer->timeout_counter)"
+				<< std::endl;
+			// Add peer to the list
+			timeouted_peers.push_back(peer->id);
+			// Don't bother going through the buffers of this one
+			continue;
+		}
+
+		float resend_timeout = dynamic_cast<UDPPeer *>(&peer)->getResendTimeout();
```

Comment:
`UDPPeer *udp_peer = dynamic_cast<UDPPeer *>(&peer);`, shortens and improves readability on 9 lines.

Reply:
exact :)